util/pull/pull.go

```diff
@@ -27,7 +27,7 @@ import (
 
 type Puller struct {
 	ContentStore content.Store
-	Resolver     *resolver.Resolver
+	Resolver     resolver.Resolver
```

Comment:
As an interim step, I created an interface in the same package that both `util/resolver/pool.Resolver struct` and `source/containerimage/OCILayoutResolver struct` satisfy. It should not be final, but lets us focus it in one place (and out of the pool resolver package), whence we can clean it up.

The only difference between what our implementations of `Resolver` do and `remotes.Resolver` are:

```go
type Resolver interface {
	remotes.Resolver
	HostsFunc(host string) ([]docker.RegistryHost, error)
	// WithSession returns a new resolver that works with new session group
	WithSession(s session.Group) Resolver

	// WithImageStore returns new resolver that can also resolve from local images store
	WithImageStore(is images.Store, mode source.ResolveMode) Resolver
}
```

I have not figured out yet where/how that gets used. If we can get around it, this whole thing does simplify to `remotes.Resolver`.

Reply:
And, it doesn't quite work, because those funcs return `*util/resolver.Resolver` (pointer to a struct) and not `util/pull.Resolver` (interface). It satisfies the interface requirements, but that doesn't make the signature identical.

Reply:
`WithSession` is used for attaching authentication. Do not need it as you are not providing credentials for accessing OCI layout.

Reply:
That should make it easier. But it is called [here](https://github.com/moby/buildkit/blob/master/util/pull/pull.go#L199), from within `util/pull/pull.go`, as part of [PullManifests()](https://github.com/moby/buildkit/blob/master/util/pull/pull.go#L108).

Actually, it is [used inside a func](https://github.com/moby/buildkit/blob/185c556282951a0edc005b99b95badeb0a022577/util/pull/pull.go#L198-L207) as a property of `PulledManifests`:

```go
	return &PulledManifests{
		Ref:              p.ref,
		MainManifestDesc: p.desc,
		ConfigDesc:       p.configDesc,
		Nonlayers:        p.nonlayers,
		Descriptors:      p.layers,
		Provider: func(g session.Group) content.Provider {
			return &provider{puller: p, resolver: p.Resolver.WithSession(g)}
		},
	}, nil
```

I cannot quite figure out where that is consumed. Either way, is it not easier to just have a `WithSession()` where the remote registry puller cares, and the local registry ignores it?

Reply:
You should be able to get the provider directly from content. You might need to pass in "registry-mode" or smth like this so you know which type to use.

Reply:
> get the provider directly from content

I don't understand.

Reply:
When you have an implementation that implements `Resolver` (or `Fetcher` or `Provider`) based on client side contents that is all you need to get `content.Provider` out of it.

There is an additional question in here about do we want the lazy pull behavior for these layers. For registry images we avoid pulling the layers until the content is really needed(this is why there is this callback here). For OCI case I think the lazy behavior doesn't make quite sense and may cause more harm than benefits. Still I think this pullmanifest may remain as is, with just a different `content.Provider` callback implementation but we might want to just call `Unlazy()` for this reference directly before returning from `Snapshot()` @sipsma